### ISSUE TITLE
Multi-queue proposal with explicit transfers

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#1073]
# Multiqueue proposal

This is a separate issue from #1065 to not pollute the investigation issue with all the discussions that can happen on the proposal. Please read the investigation for motivation and background on multi-queue in all APIs. Also see #1066 that's a different multi-queue proposal with implicit transfers. (unfortunately this proposal comes a bit later because I waited for internal reviews)

A complete multiqueue proposal for WebGPU needs to have all of these sub-proposals:

 - Discovery and requesting mechanisms for queues, if any.
 - Synchronization mechanism between queues.
 - Semantics for resources sharing / transfer between queues that prevents data-races.
 - Mechanisms / thoughts on how resources are allowed to be used on a queue or another.
 - How non-resource concepts relate to queues (in particular command encoders, but also maybe buffer mapping and pipelines)

## Synchronization proposal.

The hardest part to come up with a proposal for is the synchronization. That's because the execution of commands on queues need to follow a DAG, and we need to guarantee that if a resource is used as writeable in node W and used (read or write) in node U, then there must be a path from node W to U or from node U to W (edges are execution dependencies and prevent the execution from overlapping).

Both D3D12 and Vulkan have concepts of exclusive and concurrent access, where exclusive access means a single queue can access the resource at a time, while in concurrent access all queues can access it (but data races still disallowed).

### Proposal for exclusive access

Exclusive access synchronization is a bit simpler because there's a single owner at a time, so let's start with it. Vulkan requires a handshake between the two queues transferring ownership, so WebGPU needs the transfer command to be started from the giving queue. The simplest version would be an API like the following:

```webidl
partial interface GPUQueue {
    void transferOwnership(GPUBuffer | GPUTexture resource, GPUQueue receivingQueue);
};
```

Each resource would have an internal reference to the queue that currently owns it. `transferOwnership` would update that reference, and ensure proper synchronization happens on the giving and receiving queues. Of course a resource would only be allowed to be used on the `GPUQueue` that currently owns it.

The problem is that there might be many resources going back and forth between queues, and for the developer, it isn't very apparent what parallelism will be possible. Ideally we want the DAG of operations to be very apparent to them so they can reason about it.

The key idea is to tie resource transfer to fence signal values. (remember that a `GPUFence` can only be signaled on the queue it was created from), and to avoid having tons of fence signals, we batch ownership transfers.

```webidl
dictionary GPUResourceOwnershipTransfer {
    required GPUQueue receivingQueue;
    required sequence<GPUBuffer | GPUTexture> resources;
}

partial interface GPUQueue {
    void transferOwnership(GPUFence fence,
                           unsigned long long signalValue,
                           sequence<GPUResourceOwnershipTransfer> transfers);
                           
    void wait(GPUFence fence, unsigned long long waitValue);
};
```

In this version `GPUQueue.transferOwnership` acts like a `GPUQueue.signal` (with the same validation). But also stores the `transfers` on the `fence` for `signalValue`. It puts all resources in the `BeingTransferred` state (where they aren't owned on any queue).

`GPUQueue.wait` has the usual validation that the fence must not be an error and the `waitValue` must be less than the last value signaled. It also finishes the transfer of all the resources that where stored for transfer on `fence` for `this` `GPUQueue` for a `signalValue <= waitValue`.

Finishing the transfer requires a `GPUQueue.wait` on the fence that received the transfer, and the following indirect transfer wouldn't work (for implementation simplicity, and clarity in the API):

```js
let fence0 = queue0.createFence();
let fence1 = queue0.createFence();

queue0.transferOwnership(fence0, 1, {queue: queue2, resources: [buffer])
queue1.wait(fence0, 1);
queue1.signal(fence1, 1);
queue2.wait(fence1, 1);

// Error, buffer is still in the "BeingTransferred" state.
queue2.submit([commandBufferUsing(buffer)])
```

### Proposal for concurrent access

Buffers are free to use in concurrent access in D3D12, and in all Vulkan drivers I saw. However textures using concurrent access can be deoptimized so they would need a flag on creation to allow concurrent access. Vulkan requires the list of all queues used for concurrent access, but it doesn't seem used in drivers I saw so in WebGPU concurrent access is for all queues (we can easily add a queue list later if needed).

```webidl
partial dictionary GPUTextureDescriptor {
    // default to false, so that by default textures are optimized.
    boolean allowConcurrentQueueAccess = false;
    
    // Undefined means the device's default queue, so that by default nothing
    // changes compared to the current WebGPU API.
    // (note that starting owned by the device is useless because it only
    // allows readonly access, and the texture is newly created).
    GPUQueue initialQueueOwner;
}

partial dictionary GPUBufferDescriptor {
    // Same as for GPUTexture but must be set to null if
    // mappedAtCreation is true.
    GPUQueue initialQueueOwner;
};
```

Then we change the `GPUResourceOwnershipTransfer` to have the `GPUQueue` be optional, a `GPUQueue` of `undefined` or `null` instead means that the ownership is transferred to the device == all queues for readonly access.

```webidl
dictionary GPUResourceOwnershipTransfer {
    GPUQueue receivingQueue;
    required sequence<GPUBuffer | GPUTexture> resources;
}
```

Each resource has an internal slot that contains the list of queues it is allowed to be used on (instead of a single owning queue), that's updated on exclusive ownership transfer on `GPUQueue.wait` as described above. When `GPUQueue.wait` waits for something that's a transfer to the device, `this` is added as an allowed queue to all the resources in that transfer.

(implementation notes: lazy clearing will need to happen on device transfer, otherwise it could cause data races when trying to clear in concurrent mode. The Vulkan layout / D3D12 state will need to be the most general readonly state/layout allowed for that texture).

Moving from concurrent ownership to exclusive ownership is done via `GPUDevice.transferOwnership`:

```webidl
partial interface GPUDevice {
    void transferOwnership(sequence<GPUResourceOwnershipTransfer> transfers);
};
```

This would cause any further operation on a queue transferred to, to wait on all previous operations on all queues (loss of parallelism, see alternatives below for potential fixes?), and the resource would be immediately owned by the queue.

#### Interaction with `GPUBuffer.unmap`.

What's the queue ownership of unmapped buffers? Buffers with `MAP_WRITE` aren't writeable on the GPU (for now?) so could go back to concurrent access, while `MAP_READ` buffers are only writeable on the GPU so would go back to being owned on a queue. `GPUBuffer.unmap` would take an optional argument:

```webidl
dictionary GPUUnmapOptions {
    GPUQueue owningQueue; // concurrent access if undefined
};
partial interface GPUBuffer {
    void unmap(GPUUnmapOptions? options);
};
```

#### Implicit synchronization

WebGPU has enough information that it could theoretically do implicit synchronization between queues when it sees they both use the same resource. This is natural to think about because WebGPU already does implicit memory synchronization for execution inside a single queue.

However the goal of multi-queue is to take advantage of high-level parallelism in application's computation graph. Making synchronization implicit means that small changes to the application's code could silently result in additional synchronization that completely serializes the command flow. Likewise when building the multi-queue application, it would be extremely hard to know if we're actually taking advantage of parallelism. The combinations of these two things would make it almost impossible to write multi-queue applications and the feature mostly useless.

#### Automatic multi-queue

Similarly to implicit synchronization, WebGPU has enough information that it could theoretically automatically schedule work on multiple queues and magically make people's code faster!

This has the same drawbacks as implicit synchronization, and would be extremely difficult to implement efficiently. Also moving tiny chunks of work to a different queue would be a deoptimization due to the synchronization overhead so the WebGPU implementation would need to guess how expensive commands are, which is extremely hard to do.

#### Have the list of queues in concurrent queues be explicit

This would change `GPUResourceOwnershipTransfer` to have a `sequence<GPUQueue>` instead of a single queue, and single ownership transfer would happen if the sequence contains a single element. It could help not wait for every single queue when doing a `GPUDevice.transferOwndership`.

#### Have "device" fences that wait for all previously submitted commands

This would allow the concurrent -> exclusive transfer to not happen immediately and instead an application could signal a fence in the `GPUDevice.transferOwnership` and `GPUQueue.wait` for it later, increasing parallelism

## Queue discovery and requesting.

This is mostly orthogonal to the synchronization proposal and any proposal of the following form could work, with no constraints on whether compute-only or copy-only queues can be created:

```webidl
partial interface GPUDevice {
    GPUQueue createQueue(GPUQueueDescriptor descriptor);
    GPUQueue defaultQueue;
};
```

Although for Vulkan it might be useful to know in advance how many queues the application would like for the device.

(note: for efficiency, would it be best to have the number of queues known on device creation? (so that maps of queue->T can just be vectors of T))

## Interaction with the command encoder.

Vulkan and D3D12 need to know which queue family commands will be encoded for, so we could add a new member for the `GPUCommandEncoderDescriptor`:

```webidl
partial dictionary GPUCommandEncoderDescriptor {
    GPUQueue executionQueue; // undefined means the default queue.
};
```

Likewise for the render bundle descriptor

**Comment by kvark** [comment]
Thank you for writing this!
Here is the first batch of high-ish level feedback:

**Explicit**. I agree that being explicit about the ownership transfer, and more importantly - queue synchronization, is good. Without that, it's difficult for the user to make sure the parallelism is achieved. Some associated complexity is fine, since nobody is forced to use multiple queues.

**Fence**. I'm not sure if requiring users to do fences, in addition to the explicit hand-off, is a good thing. After all, the implementation can figure out the fences, and the user already has the means to reason about parallelism (by the explicit hand-off).

**Concurrent**. I believe concurrent access to textures is a bag of headache. We'd have to use very defensive image layouts internally, it wouldn't work with D3D12 COPY_SRC being a separate usage, and overall the gains from this API are not clear. It sounds like it could be something to follow-up, and for now focus on concurrent buffers and non-concurrent textures.

**Transfoer**. Doing the `Queue::transferOwnership` would require a separate submission on Vulkan. It seems like it would be nicer to put it into `GPUCommandBufferDescriptor` like the "Explicit handover" section of #1066 suggests.

**Comment by Kangz** [comment]
> Fence. I'm not sure if requiring users to do fences, in addition to the explicit hand-off, is a good thing. After all, the implementation can figure out the fences, and the user already has the means to reason about parallelism (by the explicit hand-off).

While technically you can do the same, tying the transfer to a fence value means that if you transfer from queue A to B, you can still submit some work on B before synchronizing with A with the fence wait. Otherwise if the transfer is immediate, any future submit on B will wait on A.

> We'd have to use very defensive image layouts internally, it wouldn't work with D3D12 COPY_SRC being a separate usage.

My reading of the D3D12 `Multi-engine synchronization doc` is that resources in the `COMMON` state can be used as readonly by all engines at the same time, for all readonly usages. But it's not super clear and it would be nice to have confirmation. This would allow for concurrent textures.

> Transfer. Doing the Queue::transferOwnership would require a separate submission on Vulkan. It seems like it would be nicer to put it into GPUCommandBufferDescriptor like the "Explicit handover" section of #1066 suggests.

Yeah that would work, or we could expand `GPUQueue.submit`.

**Comment by kvark** [comment]
> Otherwise if the transfer is immediate, any future submit on B will wait on A.

It doesn't have to be immediate. All we'd spec in this case is that there is a queue synchronization happening between the hand-off and the next usage of the resource on the destination queue. The implementation sees which resources are used, and it doesn't have to insert a semaphore, internally, until it sees a submission that needs it.

> resources in the COMMON state can be used as readonly by all engines at the same time

Ok, here comes the ugly part. I think this is incorrect (and my investigation describes it a bit in #1065), and I'll try to build a story here to prove that:

(1) `D3D12_RESOURCE_STATE_COPY_SOURCE` is described as:
> The resource is used as the source in a copy operation. Subresources must be in this state when they are used as the source of copy operation, or a blt operation. This is a read-only state.

(2) `D3D12_RESOURCE_STATE_COMMON` is described as:
> Your application should transition to this state only for accessing a resource across different **graphics** engine types.
> The COMMON state can be used for all usages on a Copy queue using the implicit state **transitions**.

So, based on this, we can conclude that if a resource is in `COMMON` state, and you use a copy operation on it, it doesn't keep the `COMMON` state. Instead, there is an implicit state promotion to `D3D12_RESOURCE_STATE_COPY_SOURCE` that's happening.

Now let's consider the multi-queue:
> Resource access (refer to Direct3D 12_RESOURCE_STATES) is determined by queue type class not queue object. There are two type classes of queue: Compute/3D queue is one type class, Copy is a second type class.
> The resource states that are shared between the two type classes (COPY_SOURCE and COPY_DEST) are considered different states for each type class.

Generally, since we are talking about textures with `ALLOW_SIMULTANEOUS_ACCESS`, their `COMMON` state is implicitly promoted to `GENERIC_READ` on the graphics queue. However, `GENERIC_READ` does *not* include the `COPY_SOURCE` of the "copy" engine class.... So my reading is that, if you start using a resource in `COMMON` on both a graphics and a copy engines, the copy engine would do an implicit transition to `COPY_SOURCE` (of engine class "copy"), while the graphics engine would treat it as `GENERIC_READ` (of engine class "graphics/compute"), and these are different states, so this is not a valid use of the API.

**Comment by Kangz** [comment]
> All we'd spec in this case is that there is a queue synchronization happening between the hand-off and the next usage of the resource on the destination queue.

That loses a bit on the explicitness of the synchronization. The original proposal for exclusive transfers has all synchronization required done by the application via fences so it's quite clear to the developer what parallelism can happen.

> [Investigation about COPY_SOURCE on copy engine]

I think it makes sense, but the documentation is very unclear, and I'm not able to understand where that restriction would be coming from in hardware. @RafaelCintron and @damyanp can you help get clarifications from the D3D12 team on whether a subresource transitioned to the `COMMON` state that was created with the `ALLOW_SIMULTANEOUS_ACCESS` flag can be used as at the same time as `COPY_SOURCE` on a copy engine and any readonly usage on compute / render engines?

**Comment by damyanp** [comment]
There's some more documentation of this here, in case you've not seen it yet: https://devblogs.microsoft.com/directx/a-look-inside-d3d12-resource-state-barriers/

Key part here is that reads from simultaneous access resources "effectively promote" the resource state from COMMON, but don’t really change anything.    So any read-only access to one of these resources is ok from multiple queues as long as the last queue to write to the resource transitions it to COMMON (either explicitly, or through implicit decay) before other queues try and read from it.

**Comment by Kangz** [comment]
Thanks for the clarification! The link you posted is incredibly useful, I wish we knew about it a couple years ago when we had to reverse engineer what things meant for optimizing barriers in Dawn. This [comment](https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/dawn_native/d3d12/TextureD3D12.cpp;l=621?q=TextureD3D12) basically describes the same thing as the blog post xD

@kvark this means that readonly concurrent access shouldn't cause issues with the copy engine, so it seems feasible!

**Comment by kvark** [comment]
I guess I'm totally confused about the queue classes in D3D12 then. Looking at this piece - https://docs.microsoft.com/en-us/windows/win32/direct3d12/user-mode-heap-synchronization#multi-queue-resource-access

I assume this section describes the resources that can be simultaneously accessed by multiple queues (i.e. buffers and `ALLOW_SIMULTANEOUS_ACCESS` textures) since otherwise the definition of the flag would be misleading:
> Allows a resource to be simultaneously accessed by multiple different queues, devices or processes (...)
> Simultaneous access allows multiple readers and one writer, as long as the writer doesn't concurrently modify the texels that other readers are accessing.

It specifically states that `COPY_SRC` on a copy engine is not the same as `COPY_SRC` on a compute/graphics engine. What's the point of this distinction in the context of  simultaneously used textures?

I'm trying to understand why the queue classes are there in the first place. Must be something that I'm missing here, it doesn't make sense yet :/

**Comment by billkris-ms** [comment]
On discrete hardware, copy engines are typically implemented using DMA between the CPU-side host and the GPU-side device.  However, graphics engines frequently implement copies by raster operations from one GPU-memory surface to another GPU-memory surface.  As such, a transition to COPY_SRC or COPY_DEST may or may not involve layout changes on the graphics queue.  On the other hand, all resources used in copy queues are required to be in the COMMON state.

From the perspective of simultaneous-access resources, this is probably not that significant since they do not have unique layout differences between read and write states.  However, since simultaneous-access resources are always in the COMMON state at the start and end of queue execution, it is totally valid to reference them on copy and graphics queues simultaneously without the need for explicit state transitions.

**Comment by kvark** [comment]
Jesse Natalie kindly clarified my confusion on D3D12 discord. Basically, the [doc section](https://docs.microsoft.com/en-us/windows/win32/direct3d12/user-mode-heap-synchronization#multi-queue-resource-access) implies the "simultaneous use" for the last paragraph (only!). As a consequence, the queue class separation totally doesn't matter for "simultaneous" resources. Therefore, my original concern about "Concurrent" being a bag of headache is invalid, we can dismiss it. I still believe we shouldn't roll out concurrent textures in the base version of multi-queue support, because there are clear performance implications, and it's not necessary to get the parallelism.

**Comment by litherum** [comment]
Is my reading correct that there are 3 subtly-different `transferOwnership()` functions in this proposal?

```
partial interface GPUQueue {
    void transferOwnership(GPUBuffer | GPUTexture resource, GPUQueue receivingQueue);
};
```

```
partial interface GPUQueue {
    void transferOwnership(GPUFence fence,
                           unsigned long long signalValue,
                           sequence<GPUResourceOwnershipTransfer> transfers);
    ...
};
```

```
partial interface GPUDevice {
    void transferOwnership(sequence<GPUResourceOwnershipTransfer> transfers);
};
```

**Comment by Kangz** [comment]
Ah the proposal text wasn't clear that the first one is just an intermediate step. I edited the text, but the only two `transferOwnership` functions are these:

```webidl
partial interface GPUQueue {
    void transferOwnership(GPUFence fence,
                           unsigned long long signalValue,
                           sequence<GPUResourceOwnershipTransfer> transfers);
};
partial interface GPUDevice {
    void transferOwnership(sequence<GPUResourceOwnershipTransfer> transfers);
};
```
